src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+#include "AddAtomicMutex.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::set;
+using std::string;
+
+namespace {
+
+class FindLoad : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+
+    set<string> symbols;
+    bool found = false;
+
+    void visit(const Load *op) override {
+        if (symbols.find(op->name) != symbols.end()) {
+            found = true;
+        }
+        include(op->predicate);
+        include(op->index);
+    }
+
+    bool find_load(Expr e, const set<string> &symbols) {
+        found = false;
+        this->symbols = symbols;
+        include(e);
+        return found;
+    }
+};
+
+class FindAtomicLetBindings : public IRGraphVisitor {
+public:
+    using IRVisitor::visit;
+
+    FindAtomicLetBindings(const set<string> &store_names)
+        : store_names(store_names) {}
+
+    void visit(const Let *op) override {
+        include(op->value);
+        let_bindings.push(op->name, op->value);
```

Comment:
Instead of push/pop pairs, use ScopedBinding:
```
include(op->value);
{
  ScopedBinding<Expr> bind(let_bindings, op->name, op->value);
  include(op->body);
}
```

Reply:
Done.